### PR TITLE
ref(MEP): Add new option to query tag values as strings from clickhouse [INGEST-1396]

### PR DESCRIPTION
**Pull request by untitaker**
For the launch of metrics-enhanced performance, there is still the risk that tag values end up being too high cardinality and too much for the string indexer to handle. For that eventuality, there will be a secondary column in Snuba that stores tags as a `Mapping[int, str]` (`tags_raw[123]`), instead of `Mapping[int, int]` (`tags[123]`).

This means that any current use of `indexer.resolve` needs to be replaced with the identity function, should we want/need that query-mode.

If that happens, the product needs to be ready for it. Unfortunately we query both Snuba and the string indexer from all over the place. So I decided the best way to find all existing uses of the string indexer was to:

* patch the indexer to crash the test if it's used without the new `pytest.mark.sentry_metrics` marker being present
* apply the new marker to every test that uses the indexer, such that the test passes again
* add a new environment variable that controls whether tests run in "strings mode" or not. It blocks access to either `tags_raw` or `tags`.
* run all the marked tests under that mode, and uncover where we are converting tag values to integers (and fixing those instances)
* fixing all the remaining usages of indexer.resolve
* adding a new CI job that reruns all marked tests with that envvar set

Also I hope with this mode it will become easier to test that a product change is still compatible with that mode.

## Changes relevant to other teams, executive summary

* All tests that touch _any_ metrics dataset have to be marked with `pytest.mark.sentry_metrics`. This will cause pytest to run those tests twice: once where the new sentry option is true (and access to `tags[]` is forbidden) and once where it is false (and access to `tags_raw[]` is forbidden)
* We are waiting on a change by the TET team to land in the metrics API that will route MEP queries to the correct metrics dataset. Until then, those tests are marked with `broken_under_tags_values_as_strings` and are exempted from being run twice.

* I noticed that we are using `_send_bucket` directly in tests. That is really not a good idea, and we should make an effort to rewrite tests that do that.
* It would be nice if the discover/mep code would also use the same methods as the metrics query builder to resolve indexer values (`sentry.sentry_metrics.utils.resolve_tag_value`) instead of implementing its own wrappers around the raw indexer

Also, let me know if you want this PR split up. The diff is enormous but most of it is renames.

### CODE REVIEW

File: src/sentry/utils/pytest/fixtures.py
Comment (untitaker) [reply src/sentry/utils/pytest/fixtures.py:434]: why would it pollute the test? I don't really care when in teardown it's happening. the point is that it is enabled until the end of the test

Comment (untitaker) [reply src/sentry/utils/pytest/fixtures.py:434]: ok, i think i get it. still, i need a way to have this setting enabled until the end of the test. if the un-patching runs after other teardown fns that's even desired, if it runs before it's not a problem either.

Comment (asottile-sentry) [reply src/sentry/utils/pytest/fixtures.py:434]: yeah generally the way I explain this is that contexts are modifying global state and ~should be re-entrant (restoring state aftewards) -- any correctly state-restoring context which runs out of order will potentially restore incorrect state

image

**Comment by onewland** [comment]
FWIW I'm already working on text tag value access here: https://github.com/getsentry/snuba/pull/2914 and I'm assuming the field will be named `tags_raw` in snql

File: src/sentry/search/events/datasets/metrics.py
Comment (wmak) [reply src/sentry/search/events/datasets/metrics.py:71]: cause `resolve_value` was effectively `resolve_tag_value` since the only difference was that we wanted to store metric_ids for the WHERE clause when using `resolve_metric`

File: src/sentry/testutils/cases.py
Comment (wmak) [reply src/sentry/testutils/cases.py:1074]: Thanks for cleaning this up 🙏

**Review by wmak — APPROVED** [review]
Changes to visibility code lgtm

File: src/sentry/snuba/metrics/datasource.py
Comment (andriisoldatenko) [reply src/sentry/snuba/metrics/datasource.py:396]: yeah i agree with you, but it was here for a while.

**Review by andriisoldatenko — APPROVED** [review]

File: .github/workflows/backend.yml
Comment (asottile-sentry) [inline .github/workflows/backend.yml:119]: I've gone through and removed all these (moving it to a default value in `setup-sentry`) so it's easier to upgrade

File: .github/workflows/backend.yml
Comment (asottile-sentry) [inline .github/workflows/backend.yml:108]: if this is a required check it should get added to the `needs` for the collector job at the bottom

**Review by asottile-sentry — COMMENTED** [review]
didn't read most of this -- mostly just reviewed the GHA -- let me know if there's specific parts you'd like me to pay closer attention to!

**Comment by untitaker** [comment]
@asottile-sentry I need approval from owners-sentry-dev in order to merge, I believe. Also in case you have committed anything, you forgot to push

File: .github/workflows/backend.yml
Comment (untitaker) [reply .github/workflows/backend.yml:108]: I wouldn't want to make this a required check and block all teams from shipping in case I messed something up.

**Comment by asottile-sentry** [comment]
> @asottile-sentry ~I need approval from owners-sentry-dev in order to merge, I believe.~ Also in case you have committed anything, you forgot to push
> 
> EDIT: actually no, I can merge as-is already

yep codeowners is currently configured as a suggestion :D

**Pull request merged**
